### ISSUE TITLE
Configuration cache report should include information about tasks that use notCompatibleWithConfigurationCache(), resulting in the cache being discarded

### ISSUE THREAD

**Issue by autonomousapps** [gradle/gradle#21290]
This information would be helpful when answering questions from developers like "why was the cache entry discarded when gradle says it had no issues storing the cache?"

### Expected Behavior
Run a build with configuration cache enabled, and a task that is incompatible:
```
./gradlew myIncompatibleTask --configuration-cache
```

Gradle will say it had no issues storing the cache, but then say the cache result was discarded. This is fine. However, Gradle will also include a configuration cache report. That report should tell the reader why the cache result was discarded.

### Current Behavior

The report does not explain why the cache result was discarded.

### Context

See also https://gradle-community.slack.com/archives/C013WEPGQF9/p1658176464418579. The alternative is having an expert you can ping in Slack to ask why a given build did what it did. Not very scalable! :)

**Comment by abstratt** [comment]
Also, this comment is probably stale, as reasons reported via `notCompatibleWithConfigurationCache()` do not seem to result in problems with Supressed severity as advertised below:

```
enum class ProblemSeverity {
    Warn,
    Failure,

    /**
     * A problem produced by a task marked as [notCompatibleWithConfigurationCache][Task.notCompatibleWithConfigurationCache].
     */
    Suppressed
}
```

**Comment by bamboo** [comment]
The comment is not stale but incomplete, what it doesn't say is that `Suppressed` problems don't make into the report.

**Comment by abstratt** [comment]
Turns out the behavior has improved since when this was reported.

In 7.6: 

```
gradle incompatible --configuration-cache 
Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: incompatible

BUILD SUCCESSFUL in 367ms
Configuration cache entry discarded.
```

In 8.7:

```
gradle incompatible --configuration-cache 
Calculating task graph as no cached configuration is available for tasks: incompatible

BUILD SUCCESSFUL in 388ms
Configuration cache entry discarded because incompatible task was found: ':incompatible'.
```

This change was implemented in the context of issue #20007, in Gradle 8.0. It still does not appear in the report though. 

P.S.: Just found out that providing a null argument as a reason turns `notCompatibleWithConfigurationCache()` into a no-op. I am not sure that was intended, and it is not documented, but won't be surprised we break folks by addressing that.
